tracer/src/Datadog.Trace/LifetimeManager.cs

```diff
@@ -109,17 +95,55 @@ private void CurrentDomain_DomainUnload(object sender, EventArgs e)
 
         private void RunShutdownTasks()
         {
+            var current = SynchronizationContext.Current;
             try
             {
-                while (_shutdownHooks.TryDequeue(out var action))
+                if (current is not null)
+                {
+                    SetSynchronizationContext(null);
```

Comment:
The previous code would only call this for `Task`s (queued using `AddAsyncShutdownTask`) but not for `Action`s (queued using `AddShutdownTask`). Is it correct to always do it?

Reply:
In this case it's safe, removing the synchronization context temporarily reduces the probability of deadlocks on Sync over Async patterns. I think is cleaner to remove the synchronization context just before running all the shutdown tasks to avoid this scenario than setting it twice in each Func<> (null and restore previous context).